File: searchspinner2_lib/src/main/java/com/mishaki/searchspinner2_lib/view/BaseSpinnerAdapter.java
package com.mishaki.searchspinner2_lib.view;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;

import com.mishaki.searchspinner2_lib.utils.SearchSpinnerConstant;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by 杜壁奇<br/>
 * on 2020/08/30 22:14
 */
public abstract class BaseSpinnerAdapter<T> extends BaseAdapter {
    @NonNull
    private String searchText = SearchSpinnerConstant.StringValues.EMPTY_VALUE;
    private List<T> list = new ArrayList<>();

    private float itemHeight = 0f;
    private boolean isShowSelectColor = true;
    private int normalItemColor = SearchSpinnerConstant.Color.WHITE;
    private int selectedItemColor = SearchSpinnerConstant.Color.GRAY;
    private int selectedPosition = SearchSpinnerConstant.Dimens.INT_DEFAULT;


    @Override
    public T getItem(int position) {
        return list.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final View rootView;
        if (searchText.length() == 0) {
            rootView = getNormalView(position, convertView, parent);
        } else {
            rootView = getSearchView(position, convertView, parent, searchText);
        }
        if (isShowSelectColor) {
            if (selectedPosition == position) {
                rootView.setBackgroundColor(selectedItemColor);
            } else {
                rootView.setBackgroundColor(normalItemColor);
            }
        }
        final AbsListView.LayoutParams param;
        if (rootView.getLayoutParams() == null) {
            param = new AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, (int) itemHeight, 0);
        } else {
            param = (AbsListView.LayoutParams) rootView.getLayoutParams();
            param.height = (int) itemHeight;
        }
        rootView.setLayoutParams(param);
        return rootView;
    }

    @NonNull
    protected abstract View getNormalView(int position, @Nullable View convertView, ViewGroup parent);

    @NonNull
    protected abstract View getSearchView(int position, @Nullable View convertView, ViewGroup parent, @NonNull String searchText);

    void setItemHeight(float itemHeight) {
        if (this.itemHeight == itemHeight) {
            return;
        }
        this.itemHeight = itemHeight;
        notifyDataSetChanged();
    }

    void setList(List<T> list) {
        this.list = list;
    }

    public void setShowSelectColor(boolean showSelectColor) {
        if (this.isShowSelectColor == showSelectColor) {
            return;
        }
        isShowSelectColor = showSelectColor;
        notifyDataSetChanged();
    }

    public void setNormalItemColor(int normalItemColor) {
        if (this.normalItemColor == normalItemColor) {
            return;
        }
        this.normalItemColor = normalItemColor;
        notifyDataSetChanged();
    }

    public void setSelectedItemColor(int selectedItemColor) {
        if (this.selectedItemColor == selectedItemColor) {
            return;
        }
        this.selectedItemColor = selectedItemColor;
        notifyDataSetChanged();
    }

    public void setSelectedPosition(int selectedPosition) {
        if (this.selectedPosition == selectedPosition) {
            return;
        }
        this.selectedPosition = selectedPosition;
        notifyDataSetChanged();
    }

    public void setItemInfo(@NonNull String searchText, @NonNull List<T> list) {
        if (this.searchText.equals(searchText) && this.list == list) {
            return;
        }
        this.searchText = searchText;
        this.list = list;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/mishaki/searchspinner2/adapter/BaseSpinnerAdapter.kt
package com.mishaki.searchspinner2.adapter

import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.BaseAdapter

abstract class BaseSpinnerAdapter<T> : BaseAdapter() {
    internal var list: MutableList<T> = ArrayList<T>()
    private var searchText = ""

    //要显示PopupWindow需要知道item的高度,要知道item的高度需要显示PopupWindow,这样就产生了悖论了
    //所以只能规定item的高度
    internal var itemHeight = 0f
        set(itemHeight) {
            field = itemHeight
            notifyDataSetChanged()
        }

    var isShowSelectColor = true
        set(isShowSelectColor) {
            if (isShowSelectColor == field) {
                return
            }
            field = isShowSelectColor
            notifyDataSetChanged()
        }
    var normalItemColor = 0xffffffff.toInt()
        set(normalItemColor) {
            if (normalItemColor == field) {
                return
            }
            field = normalItemColor
            notifyDataSetChanged()
        }
    var selectedItemColor = 0xffaaaaaa.toInt()
        set(selectedItemColor) {
            if (selectedItemColor == field) {
                return
            }
            field = selectedItemColor
            notifyDataSetChanged()
        }
    var selectedPosition = -1
        set(selectedPosition) {
            if (selectedPosition == field) {
                return
            }
            field = selectedPosition
            notifyDataSetChanged()
        }

    /**
     * 在想显示通过getSearchView方法拿到的view的时候调用,如果不想,可以不调用
     */
    fun setItemInfo(searchText: String, list: MutableList<T>) {
        this.searchText = searchText
        this.list = list
        notifyDataSetChanged()
    }

    override fun getItem(position: Int): T = list[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getCount(): Int = list.size

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val rootView: View
        if (searchText.length == 0) {
            rootView = getNormalView(position, convertView, parent)
        } else {
            rootView = getSearchView(position, convertView, parent,searchText)
        }
        if (isShowSelectColor) {
            if (selectedPosition == position) {
                rootView.setBackgroundColor(selectedItemColor)
            } else {
                rootView.setBackgroundColor(normalItemColor)
            }
        }
        val param :AbsListView.LayoutParams
        if (rootView.layoutParams == null){
            param = AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,itemHeight.toInt(),0)
        }else{
            param = rootView.layoutParams as AbsListView.LayoutParams
            param.height = itemHeight.toInt()
        }
        rootView.layoutParams = param
        return rootView
    }

    protected abstract fun getNormalView(position: Int, convertView: View?, parent: ViewGroup): View
    protected abstract fun getSearchView(position: Int, convertView: View?, parent: ViewGroup,searchText:String): View
}